src/System.Private.CoreLib/shared/System/Collections/HashHelpers.cs

```diff
@@ -86,5 +89,33 @@ public static int ExpandPrime(int oldSize)
 
             return GetPrime(newSize);
         }
+
+        // Takes an index and distributes it in the range [0, range)
+        // e.g. 0 to range exclusive.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint FastRange(uint index, uint range)
+        {
+            // Sse42.IsSupported check is baked into the R2R image so while the value returned
+            // from this method will be different on both branches, it will remain stable during
+            // a process' lifetime even if it is recompiled by tiered compilation.
+            if (Sse42.IsSupported)
+            {
+                // Using fastrange from Daniel Lemire https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/
+                //
+                // While fastrange is "fair"; it expects an evenly distributed hashCode
+                // from [0,2^32) to produce a distibuted range. As many hashCodes in .NET
+                // are generated with the high entropy in the LSB we need to redistribute
+                // the hash; for this we use Crc32. These two together allow us to skip the
+                // more costly idiv instruction.
+
+                const uint CrcReflected = 0xEDB88320;
+                return (uint)((Sse42.Crc32(CrcReflected, index) * (ulong)range) >> 32);
+            }
```

Comment:
It is a neat idea. I can think about two potential problems with this:

- This will make Dictionary slower when tiered JIT is disabled, or during startup until tiered JIT kicks in. @mjsabby Are you concerned about Dictionary being slower with tiered JIT disabled?
- There may be cases today where the `div` gives a good locality for Dictionary access. This locality will disappear  with `crc32` and the overall performance will regress. I do not know what is the chance of this happening in real world workloads.

Reply:
> This will make Dictionary slower when tiered JIT is disabled, or during startup until tiered JIT kicks in.

Does the support check get cached in a static rather than always being a CPUID? Looking at the R2R asm (`call` to `jmp`) still have quite a few cycles to spare vs the `div` branch:
```asm
       call     [Sse42:get_IsSupported():bool]
       test     al, al
       je       SHORT G_M59265_IG10
       mov      eax, 0xD1FFAB1E
       crc32    eax, r12d
       mov      edx, r13d
       imul     rax, rdx
       shr      rax, 32
       mov      ecx, eax
       jmp      SHORT G_M59265_IG11

G_M59265_IG10:		;; bbWeight=0.50
       mov      eax, r12d
       xor      rdx, rdx
       div      edx:eax, r13d
       mov      ecx, edx

G_M59265_IG11:		;; bbWeight=0.50
```

Actually, can test it. Will `COMPlus_TieredCompilation=0` mean it will always use the R2R version?

Reply:
> There may be cases today where the `div` gives a good locality for Dictionary access. This locality will disappear with `crc32` and the overall performance will regress.

This should have a lower impact with `Dictionary`; than for example `Hashtable` as its only changing the locality of the smaller `int[]` array that contains the indexes and the larger `Entry[]` array would maintain the same very packed locality. But yes it would have an impact.

Probably couldn't use in in `Hashtable` due to this greater effect as it would additionally scatter the data. Although... `Hashtable` calls modulo for every collision (whereas `Dictionary` only calls it once); so might it might balance?

However; I wouldn't want to do the work to verify that 😄

Reply:
Yes, Dictionary is one of the hottest pieces of code we run and we run with tiered jitting disabled and would have a pretty negative perf impact. I would love to see this without a regression to Dictionary in the non tiered case.

Reply:
Still is an improvement for non-tiered (I assume Sse42 is available, its just tiering that's disabled?)
```
|        Method | Toolchain |      Mean | Ratio |
|-------------- |---------- |----------:|------:|
| KNucleotide_1 |      base | 375.6  ms |  1.00 |
| KNucleotide_1 |      diff | 340.6  ms |  0.91 |
| KNucleotide_9 |      base | 100.13 ms |  1.00 |
| KNucleotide_9 |      diff |  97.05 ms |  0.97 |
```
System.Collections.TryGetValueFalse_Int, Int
```
|     Method |  Toolchain | Size |     Mean | Ratio |
|----------- |----------- |----- |---------:|------:|
| Dictionary |       base |  512 | 7.033 us |  1.00 |
| Dictionary |       diff |  512 | 3.381 us |  0.48 |
```
System.Collections.TryGetValueFalse_String, String
```
|     Method |  Toolchain | Size |     Mean | Ratio |
|----------- |----------- |----- |---------:|------:|
| Dictionary |       base |  512 | 16.46 us |  1.00 |
| Dictionary |       diff |  512 | 13.72 us |  0.83 |
```
System.Collections.TryGetValueTrue_Int, Int
```
|     Method |  Toolchain | Size |     Mean | Ratio |
|----------- |----------- |----- |---------:|------:|
| Dictionary |       base |  512 | 5.965 us |  1.00 |
| Dictionary |       diff |  512 | 3.755 us |  0.63 |
```
System.Collections.TryGetValueTrue_String, String
```
|     Method |  Toolchain | Size |     Mean | Ratio |
|----------- |----------- |----- |---------:|------:|
| Dictionary |       base |  512 | 17.92 us |  1.00 |
| Dictionary |       diff |  512 | 15.09 us |  0.84 |
```

@adamsitnik does setting the `COMPlus_TieredCompilation=0` environment variable stick for the https://github.com/dotnet/performance repository or should I be passing a flag?

Reply:
> @adamsitnik does setting the COMPlus_TieredCompilation=0 environment variable stick for the https://github.com/dotnet/performance repository or should I be passing a flag?

Its a slower so I think it does?

Reply:
Suggest looking at the disasm or similar (say COMPlus_DumpJittedMethods=1) output from a crossgen'd SPC to see exactly which dictionary types get prejitted, and then measure perf for those dictionary types with tiering disabled.

When I do this I get:
```
Compiling 19003 System.Collections.Generic.Dictionary`2[__Canon,__Canon][System.__Canon,System.__Canon]::TryGetValue, IL size = 46, hsh=0xa6ab7c8c
Compiling 20544 System.Collections.Generic.Dictionary`2[EventCacheKey,EventCacheEntry][System.Runtime.InteropServices.WindowsRuntime.WindowsRuntimeMarshal+NativeOrStaticEventRegistrationImpl+EventCacheKey,System.Runtime.InteropServices.WindowsRuntime.WindowsRuntimeMarshal+NativeOrStaticEventRegistrationImpl+EventCacheEntry]::TryGetValue, IL size = 46, hsh=0x2a6e112c
Compiling 20766 System.Collections.Generic.Dictionary`2[UInt64,__Canon][System.UInt64,System.__Canon]::TryGetValue, IL size = 46, hsh=0x1d9d08c
Compiling 21091 System.Collections.Generic.Dictionary`2[__Canon,Boolean][System.__Canon,System.Boolean]::TryGetValue, IL size = 46, hsh=0xf255b5cc
Compiling 21356 System.Collections.Generic.Dictionary`2[Int64,__Canon][System.Int64,System.__Canon]::TryGetValue, IL size = 46, hsh=0xf82920cc
Compiling 21526 System.Collections.Generic.Dictionary`2[IntPtr,__Canon][System.IntPtr,System.__Canon]::TryGetValue, IL size = 46, hsh=0x8cd0066c
Compiling 21893 System.Collections.Generic.Dictionary`2[__Canon,EventRegistrationTokenList][System.__Canon,System.Runtime.InteropServices.WindowsRuntime.WindowsRuntimeMarshal+EventRegistrationTokenList]::TryGetValue, IL size = 46, hsh=0x10419e27
Compiling 22395 System.Collections.Generic.Dictionary`2[__Canon,ResourceLocator][System.__Canon,System.Resources.ResourceLocator]::TryGetValue, IL size = 46, hsh=0x64f5791b
Compiling 22577 System.Collections.Generic.Dictionary`2[Int32,__Canon][System.Int32,System.__Canon]::TryGetValue, IL size = 46, hsh=0x6f1b392c
Compiling 22719 System.Collections.Generic.Dictionary`2[__Canon,Int32][System.__Canon,System.Int32]::TryGetValue, IL size = 46, hsh=0x432ce50c
Compiling 24671 System.Collections.Generic.Dictionary`2[EventRegistrationToken,__Canon][System.Runtime.InteropServices.WindowsRuntime.EventRegistrationToken,System.__Canon]::TryGetValue, IL size = 46, hsh=0xbdae55a8
```

So the ref,ref case will be prejitted (aka(_Canon,_Canon)) so the (string,string) result above looks promising. 

(int, int) likely is still running jitted code. Try (int, string) maybe?

Reply:
> `(int, int) likely is still running jitted code.`

We have 3 month old IBC data that have known issues. Once we get up to date IBC data, you will see more Dictionary instantiations prejited for CoreLib.

Reply:
Also seems like the perf repo tests are using unique randomly distributed keys. Might be worth running some perf examples where we have less well distributed key sets?

Reply:
I would prefer this hashcode-to-bucket mapping change to be a separate PR.

Reply:
I am not quite sure why we are using the reflected polynomial of CRC-32 (`CrcReflected`) as the initial hash value for CRC-32C calculation implemented by the intrinsic.

Reply:
> Also seems like the perf repo tests are using unique randomly distributed keys. Might be worth running some perf examples where we have less well distributed key sets?

Added some test for that, and large values https://github.com/dotnet/performance/pull/938

Reply:
> Yes, Dictionary is one of the hottest pieces of code we run

Makes my top 6 (the two waits don't count as they are the main thread and blocked forever)

![image](https://user-images.githubusercontent.com/1142958/66889714-5d145280-efdb-11e9-880d-67c233a40b72.png)